Initiate Comprehensive Kuksa Requirements Analysis for Enhanced Version Development (#21)
eclipse-kuksa/kuksa-databroker: doc/kuksa_analysis/kuksa_use_cases.md

# KUKSA Use Cases

# Content
- [KUKSA Use Cases](#kuksa-use-cases)
- [Content](#content)
- [Use Cases status](#use-cases-status)
- [Use cases:](#use-cases)
  - [UC1](#uc1)
  - [UC2](#uc2)
  - [UC3](#uc3)
  - [UC4](#uc4)
  - [UC5](#uc5)
  
# Use Cases status
| Description                               | Status                       |
|-------------------------------------------|------------------------------|
| Implemented and Verified                  | 🟢                           |
| Approved, Not Yet Implemented             | 🟡                           |
| Long Term Goal                            | 🔴                           |

# Use cases:

## UC1
    Title: Signal consumber subscribes to all signal values.

    Status: 🟡

    Description: Signal Consumer subscribes on changed sensor values or actuator current values of all signals

**Primary Actor:** Signal Consumer

> **Secondary Actors:** Databroker, Provider, Vehicle Network 

Reviewer: The Databroker isn't an actor - it is passive and reacts to requests of consumers and providers. As all the use cases are related to it, imho we do not need to mention it here at all:
```suggestion
**Secondary Actors:** Provider, Vehicle Network 
```

**Priority**: High

**Preconditions:**

>  * Signal Consumer contains a valid authentication token to connect and perform calls to databroker.

Reviewer: ```suggestion
 * The consumer passes a valid authentication token to authorize having read access rights to all requested signals.
```
or is it: "... token authorizing to have read access ..."?

**Special requirements:**
 * The use case must meet the Performance
 * What if provider died?

**Assumptions:**
 * Signal Consumer subscribes to all the possible sensor values or actuator current values to test special requirements.
 * Provider porvides signals values to Databroker as fast as receives data from the Vehicle Network.

**Postconditions:**
 * A subscription is created, and Signal Consumer receives streaming of signals.

**Sequence diagram:**

> ![Signal Consumer Subscribes](../diagrams/Signal_Consumer_subscribes.png)

Reviewer: My intention for removing some API-specific details is to have the use cases independent of the implementation.
```suggestion
The following sequence shows a situation, where a consumer subscribes to a set of signals supplied by two different providers. One has already connected to the Databroker when the consumer raises its subscription, the other provider is connecting afterwards.

![Signal Consumer Subscribes](../diagrams/Signal_Consumer_subscribes.png)
```

2.	Signal Consumer subscribes to all the sensor values or current values of actuators.
3.	Databroker checks if there is a provider up and running.
    3.1 If Provider is connected
        3.1.1 Then Databroker sends a start signal command with an unique uuid per command.
        3.1.2 If the Provider recevied correctly the command
            3.1.2.1 Then the Provider returns a response containing the uuid of the previous command and True.
            3.1.2.1 Else the Provider returns a response containing the uuid of the previous command and False.
        3.1.3 The Provider starts sending a stream of values containing all the new signal values requested.
        3.1.4 The Databroker receives the new signal values and update its database.
        3.1.5 The Databroker sends the updated values stream back to the Signal Consumer.
    3.2 if Provider is not connected.
        3.2.1 ?
4.	Signal Consumer closes subscription to Databroker.
    4.1 The Databroker sends a stop signal command with the previous uuid.
    4.2 The provider stops reading signal values from the Vehicle Network.
5.	Use case finishes.

**Exceptions:**

## UC2

>     Title: Signal Consumer actuates on an actuator.

Reviewer: ```suggestion
    Title: The consumer wants to set the value of an actuator signal.
```
